src/check/parse/tokenize/src/main.zig

```diff
@@ -0,0 +1,1399 @@
+            };
+            self.message_count += 1;
+        }
+    }
+
+    /// Returns the current byte, or null if at the end.
+    pub fn peek(self: *Cursor) ?u8 {
+        if (self.pos < self.buf.len) {
+            return self.buf[self.pos];
+        }
+        return null;
+    }
+
+    /// Returns the byte at the given lookahead offset.
+    pub fn peekAt(self: *Cursor, lookahead: usize) ?u8 {
+        if (self.pos + lookahead < self.buf.len) {
+            return self.buf[self.pos + lookahead];
+        }
+        return null;
+    }
+
+    pub fn isPeekedCharInRange(self: *Cursor, lookahead: usize, start: u8, end: u8) bool {
+        if (self.pos + lookahead < self.buf.len) {
+            const ch = self.buf[self.pos + lookahead];
+            return ch >= start and ch <= end;
+        }
+        return false;
+    }
+
+    /// Requires that the next byte is `ch`, otherwise pushes a message.
+    pub fn require(self: *Cursor, ch: u8, kind: MessageKind) void {
+        if (self.peek() == ch) {
+            self.pos += 1;
+        } else {
+            self.pushMessageHere(kind);
+        }
+    }
+
+    /// Chomps “trivia” (whitespace, comments, etc.) and returns an optional indent.
+    pub fn chompTrivia(self: *Cursor) ?Indent {
+        var sawNewline = false;
+        var indent = Indent.init();
+
+        while (self.pos < self.buf.len) {
+            const b = self.buf[self.pos];
+            if (b == ' ') {
+                self.pos += 1;
+                if (sawNewline) indent.num_spaces += 1;
+            } else if (b == '\t') {
+                self.pos += 1;
+                if (sawNewline) indent.num_tabs += 1;
+            } else if (b == '\n') {
+                self.pos += 1;
+                sawNewline = true;
+                indent = Indent.init();
+            } else if (b == '\r') {
+                self.pos += 1;
+                sawNewline = true;
+                indent = Indent.init();
+                if (self.pos < self.buf.len and self.buf[self.pos] == '\n') {
+                    self.pos += 1;
+                } else {
+                    self.pushMessageHere(.MisplacedCarriageReturn);
+                }
+            } else if (b == '#') {
+                self.pos += 1;
+                while (self.pos < self.buf.len and self.buf[self.pos] != '\n') {
+                    self.pos += 1;
+                }
+            } else if (b >= 0 and b <= 31) {
+                self.pushMessageHere(.AsciiControl);
+                self.pos += 1;
+            } else {
+                break;
+            }
+        }
+        if (sawNewline) return indent;
+        return null;
+    }
+
+    fn maybeMessageForUppercaseBase(self: *Cursor, b: u8) void {
+        if (b == 'X' or b == 'O' or b == 'B') {
+            self.pushMessageHere(.UppercaseBase);
+        }
+    }
+
+    pub fn peekUnicode(self: *Cursor) UnicodeInfo {
+        const len3 = std.unicode.utf8ByteSequenceLength(self.buf[self.pos]) catch {
```

Comment:
General question, should you be using `self.peek` instead of direct array access in many of these positions?

Peek checks to ensure you are in range this function misses that. I wonder if consistently calling peak would help avoid bugs by forcing you to check for the end of array case.